lib/sync.api.js

```diff
@@ -0,0 +1,337 @@
+   * @param {string} [buildDirectory=Aquifer.config.build.directory] Build
+   *   directory relative to the project root. Defaults to build directory
+   *   defined in aquifer.json.
+   * @param {boolean} [forceCopy=false] Set true to force copying over
+   *   symlinking for all items.
+   * @param {string} hook Name of the hook being called. Optional.
+   * @param {string} type Type of items to sync: 'file' or 'dir'. Optional.
+   * @returns {object} promise object.
+   */
+  execute(buildDirectory, forceCopy, hook, type) {
+    buildDirectory = typeof buildDirectory !== 'undefined' ? buildDirectory : this.config.build.directory;
+    forceCopy = typeof forceCopy !== 'undefined' ? forceCopy : false;
+    hook = typeof hook !== 'undefined' ? hook : false;
+    type = typeof type !== 'undefined' ? type : false;
+
+    let sync = this;
+    let itemsArr = this.getItems(hook, type);
+    let aquiferConsole = this.aquifer.console;
+
+    aquiferConsole.log('Starting ' + (hook ? hook + ' ' : '') + 'sync...');
+
+    // Process items.
+    let itemPromises = itemsArr.map(
+      function (item) {
+        return sync.executeItem(item, buildDirectory, forceCopy);
+      }
+    );
+
+    return Promise.all(itemPromises)
+      .then(function () {
+        // Reset items and excluded items.
+        sync.reset();
+        aquiferConsole.log((hook ? hook + ' s' : 'S') + 'ync complete.');
+      })
+      .catch(function (error) {
+        sync.reset();
+        aquiferConsole.log(error, 'error');
+      });
+  }
+
+  /**
+   * Executes syncing of a single item.
+   * @param {object} item The item to sync.
+   * @param {string} buildDirectory Build directory relative to the project
+   *   root. Defaults to build directory defined in aquifer.json.
+   * @param {boolean} forceCopy Set true to force copying over symlinking for
+   *   all items. Defaults to false.
+   * @returns {object} promise object.
+   */
+  executeItem(item, buildDirectory, forceCopy) {
+    buildDirectory = typeof buildDirectory !== 'undefined' ? buildDirectory : this.config.build.directory;
+    forceCopy = typeof forceCopy !== 'undefined' ? forceCopy : true;
+
+    return new Promise((resolve, reject) => {
+      let aquiferConsole = this.aquifer.console;
+      item.method = typeof item.method !== 'undefined' ? item.method : 'copy';
+      item.conflict = typeof item.conflict !== 'undefined' ? item.conflict : 'overwrite';
+      item.required = typeof item.required !== 'undefined' ? item.required : false;
+
+      // Item must have a source property defined.
+      if (typeof item.source === 'undefined' || !item.source) {
+        reject('Item with destination ' + item.destination + ' has no source property defined. Skipping.');
+        return;
+      }
+
+      // If destination is the Drupal root and the source is a directory
+      // assume we want to copy/sync the contents. Iterate over contents
+      // add them to the links array individually.
```

Comment:
We probably will need to make sure this is clear to everyone as this sounds like an interesting one off assumption. I see this used for "root"... but should we make it to where this can be the case for any directory? as opposed to just making a symlink to a directory. Maybe  a property like 'files-only` or something indicating that the directory isn't created.... but the files in that directory are moved... so you don't have to do other directories 1 at a time...